### ISSUE TITLE
Segmentation fault in restore_chartab

### ISSUE THREAD

**Issue by againxx** [neovim/neovim#14631]
- `nvim --version`:  NVIM v0.5.0-dev+1353-g0a653f7ab
- Operating system/version: Ubuntu 18.04.1
- Terminal name/version: alacritty 0.8.0-dev (fc87aaa)
- `$TERM`: tmux-256color

I met a similar segmentation fault like in #11548, but with different plugin combinations.
When I used [asynctasks](https://github.com/skywind3000/asynctasks.vim) to run a simple javascript file, the nvim crashed after the terminal buffer was closed. I have tried a `RelWithDebInfo` version that I built from source.
It seems related to [matchup](https://github.com/andymass/vim-matchup) and `synID()`. But I haven't yet found a
way to reproduce it with minimal configuration.

[backtrace.txt](https://github.com/neovim/neovim/files/6531986/backtrace.txt)

And this is my simple javascript file
```javascript
function Student(name) {
    this.name = name;
    this.hello = function () {
        console.log('Hello, ' + this.name + '!');
    }
}

var xiaoming = new Student('xiaoming');
xiaoming.hello();

console.log(xiaoming.constructor)
console.log(Student.prototype.constructor)
```

**Comment by resolritter** [comment]
`restore_chartab` assumes that `syn_win`, `syn_win->w_s` and `syn_buf` all have valid pointers

https://github.com/neovim/neovim/blob/00effff56944d5b59440dcdb5e3496d49a76d3e2/src/nvim/syntax.c#L809-L814

While that maybe works as intended for `syntax.c`'s particular lifecycle, the problem is that it ends up being invoked from the `f_syn*` public API functions in src/nvim/eval/funcs.c, which blindly (and, as showcased by this ticket, wrongly) assume that it's fine to call any `syn_*` functions at any time.

As a possible fixes I think either you protect the public API functions with `NULL` checks, like so:

```c
bool syn_ready(void) {
  return syn_win != NULL && syn_win->w_s != NULL && syn_buf != NULL;
}

static void f_synconcealed(...) {
   if (!syn_ready()) {
      // return default value
   }
}
```

Or add the `NULL` checks to `restore_chartab` if the invariant of `syn_win`, `syn_win->w_s` and `syn_buf` all being valid at all times doesn't hold.

In any case, for the reasons explained above, I expect that this bug is also present in Vim. Maybe it's a good idea to reproduce it there and later Neovim could get a patch for this if/once the fix is made.

**Comment by zeertzjq** [comment]
May be fixed by #18136

**Comment by zeertzjq** [comment]
Does this crash still exist?
